src/gt4py/analysis/passes.py

```diff
@@ -495,6 +498,55 @@ def apply(cls, transform_data: TransformData) -> None:
         transform_data.blocks = cls.SplitBlocksVisitor().visit(transform_data)
 
 
+class FieldDependencyGraphCreator(gt_ir.IRNodeVisitor):
+    @classmethod
+    def apply(cls, node_or_iterable):
+        return cls()(node_or_iterable)
+
+    def _reset_refs(self):
+        self.field_refs = {}
+        self.var_refs = []
+
+    def __call__(self, node_or_iterable):
+        self.graph = nx.DiGraph()
+        self._reset_refs()
+        self.stmt_index = 0
+        self.visit(node_or_iterable)
+        return self.graph
+
+    def visit_FieldRef(self, node: gt_ir.FieldRef):
+        if node.name not in self.graph.nodes:
+            self.graph.add_node(node.name)
+
+        if node.name in self.field_refs:
+            self.field_refs[node.name].append(node.offset)
+        else:
+            self.field_refs[node.name] = [node.offset]
+
+    def visit_VarRef(self, node: gt_ir.VarRef):
+        if node.name not in self.graph.nodes:
+            self.graph.add_node(node.name)
+
+        self.var_refs.append(node.name)
+
+    def visit_Assign(self, node: gt_ir.Assign):
+        target_name = node.target.name
+
+        self._reset_refs()
+        self.visit(node.value)
+
+        for value_field, offsets in self.field_refs.items():
+            self.graph.add_edge(target_name, value_field, offsets=offsets, index=self.stmt_index)
+
+        for value_ref in self.var_refs:
+            self.graph.add_edge(target_name, value_ref, index=self.stmt_index)
+
+        self.stmt_index += 1
```

Comment:
This logic depends on the visitor visiting the assignments in the order they appear. I'm not sure that's a requirement of the IRVisitor, but it is how it has been written.

Reply:
While it's nowhere explicitly documented, I think this is a crucial assumption for many of the visitors and passes. I wondered that too at some point, so far it's never not worked :sweat_smile: